upgrade-and-test-plan.md

```diff
@@ -0,0 +1,68 @@
+## Steps to take to upgrade and test the library
+
+The proposal for introducing the following step is to adopt a git strategy as follows:
+
+1. a `dev` branch to merge against under the `hashicorp/next-remote-watch` repo.
+2. Every step of the following plan will be introduced as a separate PR to be merged against `dev` from our fork.
+3. Once all PRs under `dev` are merged, `dev` can be merged against `master/main` in the main repo.
+
+### Upgrade dependencies
+
+Currently test dependencies include version ^9 of Next.js that exposes a different API for hot reloading (`app.hotReloader` instead of `app.server.hotReloader` introduced in Next >10).
+
+1. Upgrade `next` to latest version
+2. Install `jest`
+3. Install `eslint`
+
+_Reminder: we will need to audit, and potentially fix, high/critical vulnerabilities in third party dependencies_
+
+---
+
+### Setup ESLint and Jest
+
+Improve code quality in the project by adding a basic ESLint configuration (consider `eslint-plugin-node`).
+
+1. install `eslint`
+2. install `eslint-plugin-prettier` and `eslint-plugin-node`
+3. extend eslint recommended, prettier recommended, and node recommendede configuration.
+
+Introduce `jest` in the project:
+
+1. install `jest`
+
+---
+
+### Testing Strategy
+
+#### Step 1 - Unit Test the `main.js`
+
+1. We extract the code from the `.bin/next-remote-watch` into a `main.js` under `src/`. The `main.js` will export a `run` function that is invoked in the `.bin/next-remote-watch`
+2. We make sure that the logic in `main.js` is encapsulated into functions, which gives us the possibility to export those functions and test them individually (and mock the dependencies)
+3. We write the unit tests
+
+#### Step 2 - create modules
+
+Once we have test coverage in place, we can start refactor the `main.js` and extract each functionalities into separate modules (and extract the corresponding tests from the `main.test.js`).
+The `src/` directory will look like the following:
+
+```sh
+src/
+├── app.js
+├── main.js
+├── program.js
+├── server.js
+└── watcher.js
+```
+
+### Step 3 - Integration testing to cover different Next.js version
+
+Introduce integration testing to ensure compatibility between v9, v10 (and potentially future versions).
```

Comment:
I don't think we need to cover this case. When a new version of next.js is out, if we needed to change this library to accommodate, we'd cut a major release, add release notes making it clear that the latest version is intended to be used with X version of next.js, and move on. If anyone wanted to stick with an older version of next.js, they can pin to the older version of `next-remote-watch` until they have upgraded.

Reply:
How would you fill if we support different version of Next?

Until there is a breaking change with the dev server, the only different would be around the `app.server.hotReloader` vs `app.hotReloader`

Instead of calling the `.send` method directly from the instance of `app`, we could generate a `messanger` with a `send` property, and within that function we could check for support of different version?

Reply:
I guess what I'm proposing is that any given version of `next-remote-watch` only supports one version range. If next.js makes changes that require `next-remote-watch` to need to change internals, we'd just make those changes and run a major release. We would not maintain backwards compatibility with past versions of next.js.